### ISSUE TITLE
RCPurchaserInfo parseDatesIn:withLabel crash (1 User 6 times)

### ISSUE THREAD

**Issue by zsbee** [RevenueCat/purchases-ios#397]
**Description**
Looks like one of our users in Taiwan crashed 6 times on Mac. Do you know about a date access issue? It is not a wide-spread issue, but we would like to understand why this user got into a crash loop on their mac and how we should address this. Would updating the pod Fix this for example? This user is in their 14 day trial currently (active but cancelled subscription).

1. Environment
    1. Platform: Mac
    2. SDK version: 3.5.2
    3. OS version: macOS 14.2 (Big Sur)
    4. Xcode version: 12.1
    5. How widespread is the issue: 1 User but seems like a crash-loop, happend to them 6 times.
3. **Other information**
![image](https://user-images.githubusercontent.com/3268832/99366160-86777000-28b8-11eb-9373-913a315f2fa1.png)

```
CoreFoundation
0x00007fff203be000 + 1029791
libobjc.A.dylib
0x00007fff201de000 + 78793
CoreFoundation
0x00007fff203be000 + 1563781
CoreFoundation
0x00007fff203be000 + 405613
CoreFoundation
0x00007fff203be000 + 404008
Purchases
-[RCPurchaserInfo parseDatesIn:withLabel:] RCPurchaserInfo.m:143
Purchases
-[RCPurchaserInfo initializePurchasesAndEntitlementsWithSubscriberData:subscriptions:] RCPurchaserInfo.m:103
Purchases
-[RCPurchaserInfo configureWithSubscriberData:subscriptions:] RCPurchaserInfo.m:61
Purchases
-[RCPurchaserInfo initWithData:] RCPurchaserInfo.m:55
Purchases
-[RCBackend handle:withResponse:error:completion:] RCBackend.m:79
Purchases
-[RCBackend getSubscriberDataWithAppUserID:completion:] RCBackend.m:232
Purchases
-[RCHTTPClient handleResponse:data:error:request:completionHandler:] RCHTTPClient.m:91
Purchases
-[RCHTTPClient performRequest:path:body:headers:completionHandler:] RCHTTPClient.m:56
CFNetwork
0x00007fff24677000 + 806510
CFNetwork
0x00007fff24677000 + 127416
libdispatch.dylib
0x00007fff20199000 + 9693
libdispatch.dylib
0x00007fff20199000 + 14279
libdispatch.dylib
0x00007fff20199000 + 38398
libdispatch.dylib
0x00007fff20199000 + 41214
libdispatch.dylib
0x00007fff20199000 + 80989
libsystem_pthread.dylib
0x00007fff20341000 + 13465
libsystem_pthread.dylib
0x00007fff20341000 + 9319
```

**Comment by aboedo** [comment]
hi! 👋 
Thanks for reporting. 
I wasn't aware of this issue, I'm taking a look to see if I can figure out what's going on. Although it's always a good idea to upgrade the SDK to get the latest bug fixes, I don't believe any of the bug fixes between `3.5.2` and the current latest `3.7.6` would fix this. 

I'll update the ticket when I know more.

**Comment by zsbee** [comment]
Thanks, additional note: We see a few users trying to hack our subscription flow, all attempts were from china. We are using our own backend to verify subscriptions (they sometimes send up invalid receipts, empty receipts etc.) 
I can imagine that they might try messing with their device date as well... Not sure if that can cause such issues in your codebase, just wanted to mention.

**Comment by aboedo** [comment]
there's a chance that they're manually editing `NSUserDefaults` and removing their `expirationDate` there, in an attempt to extend their subscription duration. That would result in this crash

**Comment by zsbee** [comment]
Can it be the case that the paying users revenuecat user id changed sometime? in this case it looks like the userdefaults will be saved to a different key so that might also cause this?

**Comment by aboedo** [comment]
the appUserID could change if you use `identify`, for example. However, the value stored as purchaserInfo in `NSUserDefaults` should still contain an expiration date for all transactions for any user that had transactions. From the logs above, it looks like at least one of the transactions was missing an expiration date. 
Modifying the device date would also not cause this crash, since the issue is a missing value within the purchaser info.

**Comment by aboedo** [comment]
Tentatively closing out this ticket due to inactivity. @zsbee if you're still facing issues feel free to reopen

**Comment by alekplay** [comment]
Hi, I'm seeing the same crash on my personal device when running a development build. I'm calling `purchaserInfoWithCompletionBlock:` which in turn crashes on `parseDatesIn:withLabel` and `[NSDateFormatter dateFromString:]`. 

From back when I did my own receipt validation I remember the `purchase_date` (etc.) properties being very finicky with inconsistent date formats for non-US App Store regions. I had to use `purchase_date_ms` (etc.) properties instead. Unlikely this is your problem, but just throwing it out there.
